### ISSUE TITLE
[Performance] GetRootGameObjects improvements

### ISSUE THREAD

**Issue by legoritma** [modesttree/Zenject#37]
Currently due to an older bug, root objects are gotten with workaround. After 2017.2 scene.GetRootGameObjects is working. [Reference forum post](https://forum.unity.com/threads/scene-getrootgameobjects-throw-exception-even-when-object-is-loaded-bug.491877/#post-3211883)

We tested on our game 2017.4.30 and 2018.4.10 it's working perfectly fine. Can we add something like that?

```
#if UNITY_2017_2_OR_NEWER
                return scene.GetRootGameObjects()
                    .Where(x => x.GetComponent<ProjectContext>() == null);
#endif
```

**Comment by extrawurst** [comment]
Did you measure what impact that has on performance?

**Comment by svermeulen** [comment]
Are you saying that we don't need to use this code anymore?

```csharp
                return Resources.FindObjectsOfTypeAll<GameObject>()
                    .Where(x => x.transform.parent == null
                            && x.GetComponent<ProjectContext>() == null
                            && x.scene == scene);
```

That would be great since that's hacky and less performant than scene.GetRootGameObjects.  But according to the comment in ZenUtilInternal.GetRootGameObjects, it doesn't work when executed during Awake

**Comment by legoritma** [comment]
@svermeulen Yes exactly. I think this is the issue which is explained in the comment.[https://issuetracker.unity3d.com/issues/scene-is-not-considered-loaded-when-awake-is-called](https://issuetracker.unity3d.com/issues/scene-is-not-considered-loaded-when-awake-is-called)

@Extrawurst I wasn't remember the original ratios so tested again in our game on 2018.4.10 Editor.
Current method -> 317 KB GC, 3.95 ms
Suggested method -> 0.8 KB GC, 0.01 ms
